### PR TITLE
added the workspace symbol provider for markdown #46406

### PR DESCRIPTION
**Pull request by pradeepmurugesan**
#46406

### CODE REVIEW

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+import MDDocumentSymbolProvider from './documentSymbolProvider';
+import { Dictionary, flatMap } from 'lodash';
+
+export default class MarkdownWorkspaceSymbolProvider implements WorkspaceSymbolProvider {
+	private symbolProvider: MDDocumentSymbolProvider;
+	private symbolCache: Dictionary<SymbolInformation[]> = {};
+
+	public constructor(symbolProvider: MDDocumentSymbolProvider) {
+		this.symbolProvider = symbolProvider;
+		this.populateSymbolCache();
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts]: This should be done lazily when workspace symbols are first requested

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+			}
+		}
+	}
+
+	private async getSymbol(document: TextDocument): Promise<SymbolInformation[]> {
+		return this.symbolProvider.provideDocumentSymbols(document);
+	}
+
+	private registerOnSaveEvent(): void {
+		workspace.onDidSaveTextDocument(async document => {
```
Comment (mjbvz) [inline extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts]: This registration returns a registration handle that handles unregistering the event handler. Add a dispose method on `MarkdownWorkspaceSymbolProvider ` and then call  `.dipose()` on the the registration handle to make sure it is properly cleaned up

**Review by mjbvz — CHANGES_REQUESTED** [review]
Thanks for looking into this! Just a few changes before we can merge this in. Let me know if you have any questions about my comments

**Review by nzec — APPROVED** [review]
Awesome!

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 1 above)
Comment (pradeepmurugesan) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts]: done..

File: extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts (hunk 2 above)
Comment (pradeepmurugesan) [reply extensions/markdown-language-features/src/features/workspaceSymbolProvider.ts]: @mjbvz I have one question here. I have made the changes though..  Excuse me if my question is naive..

Who calls the dispose method? I couldn't find the same in the WorkspaceSymbolProvider interface as well..  I assume there should be a common place to check if the method "dispose" exists on an object and then invoke this. 

Can you kindly point me to that.

**Review by ahmadmiz — APPROVED** [review]

**Comment by pradeepmurugesan** [comment]
@mjbvz got a chance to look into the changes..  ?

**Comment by mjbvz** [comment]
Thanks! Will be in the next insiders build and in VS Code 1.23

**Comment by pradeepmurugesan** [comment]
@mjbvz Thanks for merging.. I had asked a question in the review comment. It will be really helpful if you could answer the same.

Who calls the dispose method? I couldn't find the same in the WorkspaceSymbolProvider interface as well.. I assume there should be a common place to check if the method "dispose" exists on an object and then invoke this.

Can you kindly point me to that.

**Comment by mjbvz** [comment]
In this case it will never get called. The extension would have to register the provider as one of its disposables but this is pointless because the extension itself never gets disposed. It's still good practice to clean up any resources in a `dispose` method though, which does come in handy today when writing tests

**Comment by mjbvz** [comment]
@pradeepmurugesan By the way, do you have a twitter handle that you would like mentioned when we call out this feature? Tweeting can help get some additional testing on the feature and I want to make sure proper credit is given for your work

**Comment by pradeepmurugesan** [comment]
@mjbvz Sure.. This is my twitter profile https://twitter.com/pradeepm_14. Handle- @pradeepm_14

**Pull request merged**